### PR TITLE
Added support for level 0 dispatching

### PR DESCRIPTION
**Pull request by makart19**

### CODE REVIEW

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 1]
```diff
@@ -181,14 +337,45 @@ class OpenClKernel : public Base, public KernelIface
 
     ExecutionTargetId getTarget() const { return _executionTarget; }
 
-    const OpenClKernelRef & getRef() const { return _clKernelRef; }
+    virtual cl::sycl::kernel toSycl(const cl::sycl::context & _ctx) const {}
```
Comment (amgrigoriev) [inline include/oneapi/internal/kernel_scheduler_sycl.h:340]: Nothing to return?

Comment (amgrigoriev) [reply include/oneapi/internal/kernel_scheduler_sycl.h:340]: If this method isn't expected to be called it should fire a error maybe?

**Comment by napetrov** [comment]
/intelci: run
set infra_branch=dev/amk_level0

File: examples/cpp_sycl/makefile_lnx [hunk 2]
```diff
@@ -78,27 +78,36 @@ ifndef TBBROOT
...
 else
     override threading = parallel
-    DAAL_LIB_T := $(DAAL_PATH)/libdaal_thread.$(RES_EXT)
+    DAAL_LIB_T := $(DAAL_LIB_EXT),-ldaal_thread
     EXT_LIB += $(addprefix -L,$(TBB_PATH)) -ltbb -ltbbmalloc
 endif
 
-DAAL_LIB := $(DAAL_PATH)/libdaal_core.$(RES_EXT) $(DAAL_LIB_T)
+DAAL_LIB := -L$(DAAL_PATH),$(DAAL_LIB_EXT),-ldaal_core,-Bdynamic -Wl,$(DAAL_LIB_T)
+#LEVEL0_LIB := libze_intel_gpu.so.0.8 $l0_lib/libze_loader.so
```
Comment (dr-pain) [inline examples/cpp_sycl/makefile_lnx:98]: remove comments

File: examples/cpp_sycl/makefile_lnx [hunk 3]
```diff
@@ -78,27 +78,36 @@ ifndef TBBROOT
...
     override threading = parallel
-    DAAL_LIB_T := $(DAAL_PATH)/libdaal_thread.$(RES_EXT)
+    DAAL_LIB_T := $(DAAL_LIB_EXT),-ldaal_thread
     EXT_LIB += $(addprefix -L,$(TBB_PATH)) -ltbb -ltbbmalloc
 endif
 
-DAAL_LIB := $(DAAL_PATH)/libdaal_core.$(RES_EXT) $(DAAL_LIB_T)
+DAAL_LIB := -L$(DAAL_PATH),$(DAAL_LIB_EXT),-ldaal_core,-Bdynamic -Wl,$(DAAL_LIB_T)
+#LEVEL0_LIB := libze_intel_gpu.so.0.8 $l0_lib/libze_loader.so 
+LEVEL0_LIB := -L$(DRIVER_PATH) -lze_loader
```
Comment (dr-pain) [inline examples/cpp_sycl/makefile_lnx:99]: DRIVER_PATH variable? Who will define it at customers level???

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 4]
```diff
@@ -102,18 +103,136 @@ DAAL_DECLARE_OPENCL_OPERATOR(cl_program, RetainProgram);
...
     {
-        _progamName            = programName;
+        initOpenClProgramRef(clContext, clDevice, programName, programSrc, options, status);
+    }
+
+    explicit OpenClProgramRef(cl_context clContext, cl_device_id clDevice, ze_device_handle_t zeDevice, const char * programName,
+                              const char * programSrc, const char * options, services::Status * status = nullptr)
+    {
+        initOpenClProgramRef(clContext, clDevice, programName, programSrc, options, status);
+        if (status != nullptr && !status->ok())
```
Comment (rlnx) [inline include/oneapi/internal/kernel_scheduler_sycl.h:225]: Same problem with status checks, consider
```
if (!localStatus.ok())
{
    services::internal::tryAssignStatus(status, localStatus);
    return;
}
```

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 5]
```diff
@@ -102,18 +103,136 @@ DAAL_DECLARE_OPENCL_OPERATOR(cl_program, RetainProgram);
...
+        {
+            return;
+        }
+        initModuleLevel0(zeDevice, status);
+    }
+
+    void initOpenClProgramRef(cl_context clContext, cl_device_id clDevice, const char * programName, const char * programSrc, const char * options,
```
Comment (rlnx) [inline include/oneapi/internal/kernel_scheduler_sycl.h:232]: Make this function `private`

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 6]
```diff
@@ -141,10 +260,34 @@ class OpenClProgramRef : public OpenClResourceRef<cl_program, OpenClRetainProgra
         DAAL_CHECK_OPENCL(err, status)
     }
 
-    const char * getName() const { return _progamName.c_str(); }
+    void initModuleLevel0(ze_device_handle_t & zeDevice, services::Status * status = nullptr)
```
Comment (rlnx) [inline include/oneapi/internal/kernel_scheduler_sycl.h:263]: Make this function `private`

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 7]
```diff
@@ -141,10 +260,34 @@ class OpenClProgramRef : public OpenClResourceRef<cl_program, OpenClRetainProgra
...
+    {
+        size_t binary_size = 0;
+        DAAL_CHECK_OPENCL(clGetProgramInfo(get(), CL_PROGRAM_BINARY_SIZES, sizeof(size_t), &binary_size, NULL), status);
+
+        auto binary = (unsigned char *)malloc(binary_size);
+        if (binary == nullptr) DAAL_CHECK_OPENCL(CL_OUT_OF_HOST_MEMORY, status);
```
Comment (rlnx) [inline include/oneapi/internal/kernel_scheduler_sycl.h:269]: Use DAAL_CHECK_MALLOC

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 8]
```diff
@@ -141,10 +260,34 @@ class OpenClProgramRef : public OpenClResourceRef<cl_program, OpenClRetainProgra
         DAAL_CHECK_OPENCL(err, status)
     }
 
-    const char * getName() const { return _progamName.c_str(); }
+    void initModuleLevel0(ze_device_handle_t & zeDevice, services::Status * status = nullptr)
+    {
+        size_t binary_size = 0;
+        DAAL_CHECK_OPENCL(clGetProgramInfo(get(), CL_PROGRAM_BINARY_SIZES, sizeof(size_t), &binary_size, NULL), status);
+
+        auto binary = (unsigned char *)malloc(binary_size);
```
Comment (rlnx) [inline include/oneapi/internal/kernel_scheduler_sycl.h:268]: Where is corresponding `free`?

File: include/oneapi/internal/execution_context_sycl.h [hunk 9]
```diff
@@ -113,20 +148,42 @@ class OpenClKernelFactory : public Base, public ClKernelFactoryIface
         }
         else
         {
-            auto kernelRef = OpenClKernelRef(_currentProgramRef->get(), kernelName, &localStatus);
-            if (!localStatus.ok())
+            if (!_deviceQueue.get_device().template get_info<sycl::info::device::opencl_c_version>().empty())
```
Comment (rlnx) [inline include/oneapi/internal/execution_context_sycl.h:151]: I suggest to reorganize it this way:
```cpp
KernelPtr kernel;
if (...) {
    ...
    kernel.reset(new OpenClKernelNative(_executionTarget, *_currentProgramRef, kernelRef));
}
else {
    ...
    kernel.reset(new OpenClKernelLevel0(_executionTarget, *_currentProgramRef, kernelRef));
}

kernelHashTable.add(key, kernel, localStatus);
if (!localStatus.ok()) {
    services::internal::tryAssignStatus(status, localStatus);
    return KernelPtr();
}
return kernel;
```

Comment (rlnx) [reply include/oneapi/internal/execution_context_sycl.h:151]: Minor comment: I suggest to not use "negative if"
```
if (!const) { A } else { B } // not very good
if (cond) { B } else { A } // good
```

Comment (rlnx) [reply include/oneapi/internal/execution_context_sycl.h:151]: It's difficult to understand the condition like `_deviceQueue.get_device().template get_info<sycl::info::device::opencl_c_version>().empty()`, it's worth to define a variable or method, like this 

```cpp
const bool isOpenCLBackendAvailable = _deviceQueue.get_device().template get_info<sycl::info::device::opencl_c_version>().empty()
```

**Comment by napetrov** [comment]
/intelci: run
set infra_branch=dev/amk_level0

**Comment by napetrov** [comment]
/intelci: run
set infra_branch=dev/amk_level0

File: include/oneapi/internal/kernel_scheduler_sycl.h (hunk 1 above)
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h]: fixed

File: include/oneapi/internal/error_handling.h [hunk 10]
```diff
@@ -101,6 +111,56 @@ inline services::ErrorPtr convertOpenClErrorToErrorPtr(cl_int clError)
     return services::Error::create(services::ErrorID::ErrorExecutionContext, services::ErrorDetailID::OpenCL, getOpenClErrorDescription(clError));
 }
 
+inline services::String getLevelZeroErrorDescription(ze_result_t zeError)
+{
+#define LEVEL_ZERO_ERROR_CASE(x) \
```
[the remark this answers is not in the document]
Comment (makart19) [reply include/oneapi/internal/error_handling.h]: fixed

File: include/oneapi/internal/kernel_scheduler_sycl.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h]: fixed

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 11]
```diff
@@ -102,18 +103,136 @@ DAAL_DECLARE_OPENCL_OPERATOR(cl_program, RetainProgram);
...
+private:
+    OpenClDeviceRef _clDeviceRef;
+};
+
+class LevelZeroOpenClInteropContext : public Base
+{
+public:
+    LevelZeroOpenClInteropContext() = default;
+
+    LevelZeroOpenClInteropContext(const LevelZeroOpenClInteropContext & other)
```
[the remark this answers is not in the document]
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h]: removed it, thank you.

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 12]
```diff
@@ -102,18 +103,136 @@ DAAL_DECLARE_OPENCL_OPERATOR(cl_program, RetainProgram);
...
+
+    void reset(ze_device_handle_t levelZeroDevice, services::Status * status = nullptr)
+    {
+        _levelZeroDevice = levelZeroDevice;
+        ze_device_properties_t levelZeDeviceProperties;
+        levelZeDeviceProperties.version = ZE_DEVICE_PROPERTIES_VERSION_CURRENT;
+        ze_result_t ret                 = zeDeviceGetProperties(levelZeroDevice, &levelZeDeviceProperties);
+
+        cl_device_id clDevice;
+        findDevice(&clDevice, levelZeDeviceProperties.vendorId, levelZeDeviceProperties.coreClockRate, status);
```
[the remark this answers is not in the document]
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h]: fixed

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 13]
```diff
@@ -102,18 +103,136 @@ DAAL_DECLARE_OPENCL_OPERATOR(cl_program, RetainProgram);
...
+        DAAL_CHECK_OPENCL(err, status)
+    }
+
+    void findDevice(cl_device_id * pClDevice, unsigned int vendor_id, unsigned int frq, services::Status * status = nullptr)
+    {
+        constexpr auto maxPlatforms = 16;
+        cl_platform_id platIds[maxPlatforms];
+        cl_uint nplat, ndev;
+
+        DAAL_CHECK_OPENCL(clGetPlatformIDs(maxPlatforms, platIds, &nplat), status);
```
[the remark this answers is not in the document]
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h:179]: fixed

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 14]
```diff
@@ -102,18 +103,136 @@ DAAL_DECLARE_OPENCL_OPERATOR(cl_program, RetainProgram);
...
+    OpenClDeviceRef getDeviceRef() { return _clDeviceRef; }
+
+private:
+    OpenClDeviceRef _clDeviceRef;
+};
+
+class LevelZeroOpenClInteropContext : public Base
+{
+public:
+    LevelZeroOpenClInteropContext() = default;
```
[the remark this answers is not in the document]
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h:141]: added initialization of this member

File: include/oneapi/internal/kernel_scheduler_sycl.h (hunk 5 above)
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h:248]: fixed

File: include/oneapi/internal/kernel_scheduler_sycl.h (hunk 6 above)
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h]: fixed

File: include/oneapi/internal/kernel_scheduler_sycl.h (hunk 7 above)
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h]: used services::internal::tryAssignStatus(status, services::ErrorMemoryAllocationFailed);
due to DAAL_CHECK_MALLOC makes return

File: include/oneapi/internal/kernel_scheduler_sycl.h (hunk 8 above)
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h:285]: added

File: include/oneapi/internal/kernel_scheduler_sycl.h [hunk 15]
```diff
@@ -102,18 +103,136 @@ DAAL_DECLARE_OPENCL_OPERATOR(cl_program, RetainProgram);
...
+        DAAL_CHECK_OPENCL(clGetPlatformIDs(maxPlatforms, platIds, &nplat), status);
+
+        for (int pidx = 0; pidx < nplat && pidx < maxPlatforms; pidx++)
+        {
+            if (clGetDeviceIDs(platIds[pidx], CL_DEVICE_TYPE_GPU, 1, pClDevice, &ndev) == CL_SUCCESS)
+            {
+                if (ndev > 0)
+                {
+                    cl_uint dVid = 0;
+                    clGetDeviceInfo(*pClDevice, CL_DEVICE_VENDOR_ID, sizeof(cl_uint), &dVid, NULL);
```
[the remark this answers is not in the document]
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h:188]: Possible return values are CL_SUCCESS, CL_INVALID_DEVICE if device is not valid (previous check confirm that device is valid).
CL_INVALID_VALUE if param_name is not one of the supported values ... (only two certain params are used here)
Therefore i don't see any profit from it's ret status checking

File: examples/cpp_sycl/makefile_lnx (hunk 2 above)
Comment (makart19) [reply examples/cpp_sycl/makefile_lnx]: fixed

File: examples/cpp_sycl/makefile_lnx (hunk 3 above)
Comment (makart19) [reply examples/cpp_sycl/makefile_lnx]: fixed

File: include/oneapi/internal/execution_context_sycl.h (hunk 9 above)
Comment (makart19) [reply include/oneapi/internal/execution_context_sycl.h]: fixed

File: include/oneapi/internal/kernel_scheduler_sycl.h (hunk 4 above)
Comment (makart19) [reply include/oneapi/internal/kernel_scheduler_sycl.h]: fixed

**Comment by napetrov** [comment]
/intelci: run
set infra_branch=dev/amk_level0

**Comment by dr-pain** [comment]
Good job @napetrov ! do changes and commit them without any review-approvals.

**Pull request merged**